zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -270,69 +270,156 @@ impl DbFormatChange {
 
         // Check if we need to prune the note commitment trees in the database.
         if older_disk_version < version_for_pruning_trees {
-            let mut height = Height(0);
-            let mut prev_sapling_tree = upgrade_db.sapling_tree_by_height(&height);
-            let mut prev_orchard_tree = upgrade_db.orchard_tree_by_height(&height);
-            let sapling_cf = upgrade_db
-                .db
-                .cf_handle("sapling_note_commitment_tree")
-                .unwrap();
-            let orchard_cf = upgrade_db
-                .db
-                .cf_handle("orchard_note_commitment_tree")
-                .unwrap();
-
-            height = height.next();
-
-            // Go through every height from genesis to the tip of the old version. If the state was
-            // downgraded, some heights might already be upgraded. (Since the upgraded format is
-            // added to the tip, the database can switch between lower and higher versions at any
-            // block.)
-            //
-            // Keep upgrading until the initial database has been upgraded.
-            while height <= initial_tip_height {
-                // Return early if the task is cancelled by a shutdown.
-                if !matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty)) {
-                    return;
-                }
-
+            // Get network upgrade heights
+            let (&sapling_height, _) = NetworkUpgrade::activation_list(network)
+                .iter()
+                .find(|(_, upgrade)| **upgrade == NetworkUpgrade::Sapling)
+                .expect("there should be sapling upgrade");
+            let (&orchard_height, _) = NetworkUpgrade::activation_list(network)
+                .iter()
+                .find(|(_, upgrade)| **upgrade == NetworkUpgrade::Nu5)
+                .expect("there should be Nu5 upgrade");
+
+            // Delete duplicates before sapling and orchard heights
+            let (mut prev_sapling_tree, mut prev_orchard_tree) = {
+                let height = Height(1);
                 let mut batch = DiskWriteBatch::new();
-                for _ in 0..1_000 {
-                    if height >= initial_tip_height {
-                        break;
-                    }
-                    let sapling_tree = upgrade_db.sapling_tree_by_height(&height);
-                    let orchard_tree = upgrade_db.orchard_tree_by_height(&height);
 
-                    if prev_sapling_tree == sapling_tree {
-                        batch.delete_sapling_tree(&sapling_cf, &height);
-                    }
+                batch.delete_range_sapling_tree(&upgrade_db, &height, &sapling_height);
+                batch.delete_range_orchard_tree(&upgrade_db, &height, &orchard_height);
```

Comment:
Should these be `sapling_height - 1` and `orchard_height - 1`?

Reply:
> I don't know how to come up with a definite answer, but I'm convinced they should.

I think they should too.

Reply:
Oh, at that time of the review, I also assumed the upper bound would be deleted. It occurred to me later on to check the bounds.